### PR TITLE
Add streaming support

### PR DESCRIPTION
**Pull request by jaymo001**
Add streaming support for anchored features.
Example:
1. Define streaming input source on Kafka
Define input data schema: 
```
 schema = AvroJsonSchema(schemaStr="""
    {
        "type": "record",
        "name": "DriverTrips",
        "fields": [
            {"name": "driver_id", "type": "long"},
            {"name": "trips_today", "type": "int"},
            {
                "name": "datetime",
                "type": {"type": "long", "logicalType": "timestamp-micros"}
            }
        ]
    }
    """)
stream_source = KafKaSource(name="kafkaStreamingSource",
                              kafkaConfig=KafkaConfig(brokers=["feathrazureci.servicebus.windows.net:9093"],
                                                      topics=["feathrcieventhub"],
                                                      schema=schema)
                              )
```
3. Define feature definition with the Kafka source
```
  driver_id = TypedKey(key_column="driver_id",
                          key_column_type=ValueType.INT64,
                          description="driver id",
                          full_name="nyc driver id")

    kafkaAnchor = FeatureAnchor(name="kafkaAnchor",
                                      source=stream_source,
                                      features=[Feature(name="f_modified_streaming_count",
                                                        feature_type=INT32,
                                                        transform="trips_today + 1",
                                                        key=driver_id),
                                                Feature(name="f_modified_streaming_count2",
                                                        feature_type=INT32,
                                                        transform="trips_today + 2",
                                                        key=driver_id)]
                                      )
```
5. Start streaming job
```
 redisSink = RedisSink(table_name="kafkaSampleDemoFeature", streaming=True, streamingTimeoutMs=10000)
    settings = MaterializationSettings(name="kafkaSampleDemo",
                                   sinks=[redisSink],
                                   feature_names=['f_modified_streaming_count']
                                   )
    client.materialize_features(settings)  # Will streaming for 10 seconds since streamingTimeoutMs is 10000
```

### CODE REVIEW

**Comment by hangfei** [comment]
Is schema required? can it work without schema like frame-offline?

**Comment by jaymo001** [comment]
> 
Yes. I believe it is. Kafka returns a bytes array of any type, so we have to rely on the schema to create a decoder.

File: README.md [hunk 1]
```diff
@@ -200,9 +200,10 @@ Follow the [quick start Jupyter Notebook](./feathr_project/feathrcli/data/feathr
 - [x] Private Preview release
 - [x] Public Preview release
 - [ ] Future release
```
Comment (hangfei) [inline README.md:202]: let's add what we already supports here so users can easily to find them?

like point-in-time join, aggregation, offline transformation, preprocessing?

File: feathr_project/feathr/client.py [hunk 2]
```diff
@@ -685,3 +690,24 @@ def _get_snowflake_config_str(self):
...
+
+    def _get_kafka_config_str(self):
+        """Construct the Kafka config string. The endpoint, access key, secret key, and other parameters can be set via
+        environment variables."""
+        kafka_endpoint = self.kafka_endpoint
+        # if kafka endpoint is set in the feathr_config, then we need other environment variables
+        # keys can't be only accessed through environment
+        access_key_name = _EnvVaraibleUtil.get_environment_variable('KAFKA_SHARED_ACCESS_KEY_NAME')
+        access_key = _EnvVaraibleUtil.get_environment_variable('KAFKA_SHARED_ACCESS_KEY')
+        username = _EnvVaraibleUtil.get_environment_variable('KAFKA_USERNAME')
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/feathr/client.py:702]: On Azure, it is username, i.e.  'sasl.username': username

File: src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala [hunk 3]
```diff
@@ -278,6 +278,149 @@ class FeatureGenIntegTest extends FeathrIntegTest {
...
+        |
+        |kafkaStreamingSource: {
+        |    type: KAFKA
+        |    config: {
+        |        brokers: ["feathrazureci.servicebus.windows.net:9093"]
+        |        topics: [feathrcieventhub]
+        |
+        |        schema: {
+        |            type = "avro"
+        |            avroJson:"\n    {\n        \"type\": \"record\",\n        \"name\": \"DriverTrips\",\n        \"fields\": [\n            {\"name\": \"driver_id\", \"type\": \"long\"},\n            {\"name\": \"trips_today\", \"type\": \"int\"},\n            {\n                \"name\": \"datetime\",\n                \"type\": {\"type\": \"long\", \"logicalType\": \"timestamp-micros\"}\n            }\n        ]\n    }\n    "
```
Comment (hangfei) [inline src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala:409]: pretty print the schema so it's easer to read?

File: src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala [hunk 4]
```diff
@@ -278,6 +278,149 @@ class FeatureGenIntegTest extends FeathrIntegTest {
...
+         |      params: {
+         |        table_name: "nycFeaturesStreaming"
+         |        streaming: true
+         |      }
+         |   }
+         |  ]
+         |}
+         |features: [f_modified_streaming_count, f_modified_streaming_count2, trips_today]
+      """.stripMargin
+    val featureDefConfig =
```
Comment (hangfei) [inline src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala:305]: identation is broken. indent properly so it's easy to read.

File: src/test/scala/com/linkedin/feathr/offline/generation/TestPushToRedisOutputProcessor.scala [hunk 5]
```diff
@@ -2,11 +2,11 @@ package com.linkedin.feathr.offline.generation
 
 import com.linkedin.feathr.common.types.protobuf.FeatureValueOuterClass
 import com.linkedin.feathr.common.{FeatureInfo, FeatureTypes, Header, TaggedFeatureName}
-import com.linkedin.feathr.offline.generation.outputProcessor.PushToRedisOutputProcessor
+import com.linkedin.feathr.offline.generation.outputProcessor.RedisOutputUtils
 import com.linkedin.feathr.offline.{AssertFeatureUtils, TestFeathr}
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.catalyst.encoders.RowEncoder
-import org.apache.spark.sql.types.{ArrayType, BooleanType, FloatType, IntegerType, StringType, StructField, StructType}
+import org.apache.spark.sql.types._
```
Comment (hangfei) [inline src/test/scala/com/linkedin/feathr/offline/generation/TestPushToRedisOutputProcessor.scala:9]: convention: avoid wildcard import?

File: src/main/scala/org/apache/spark/customized/CustomGenericRowWithSchema.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    val oos = new ObjectOutputStream(bos)
+    oos.writeObject(obj)
+    bos.toByteArray
+  }
+  override def deserialize(datum: Any): CustomGenericRowWithSchema = {
+    val bis = new ByteArrayInputStream(datum.asInstanceOf[Array[Byte]])
+    val ois = new ObjectInputStream(bis)
+    val obj = ois.readObject()
+    obj.asInstanceOf[CustomGenericRowWithSchema]
+  }
```
Comment (hangfei) [inline src/main/scala/org/apache/spark/customized/CustomGenericRowWithSchema.scala:40]: convention: avoid acronym

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  private def getKafkaAuth(ss: SparkSession): String = {
+    // If user set password, then we use password to auth
+    ss.conf.getOption(kafkaEndpoint) match {
+      case Some(_) =>
+          val accessKeyName = ss.conf.get(sharedAccessKeyName)
+          val endpoint = ss.conf.get(kafkaEndpoint)
+          val username = ss.conf.get(kafkaUsername)
+          val accessKey = ss.conf.get(sharedAccessKey)
+          val EH_SASL = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\""+username+"\"" +
+            " password=\"Endpoint="+endpoint+";SharedAccessKeyName="+accessKeyName+";SharedAccessKey="+accessKey+"\";"
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:31]: document why this is contructed in this way for future dev.

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  val sharedAccessKey = KafkaResourceInfoSetter.SHARED_ACCESS_KEY
+
+  private def getKafkaAuth(ss: SparkSession): String = {
+    // If user set password, then we use password to auth
+    ss.conf.getOption(kafkaEndpoint) match {
+      case Some(_) =>
+          val accessKeyName = ss.conf.get(sharedAccessKeyName)
+          val endpoint = ss.conf.get(kafkaEndpoint)
+          val username = ss.conf.get(kafkaUsername)
+          val accessKey = ss.conf.get(sharedAccessKey)
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:29]: what if these are not set? throw warning or exception as well?

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    _ss.readStream
+        .format(format)
+        .options(kafkaOptions)
+        .option("kafka.group.id", UUID.randomUUID().toString)
+        .option("kafka.sasl.jaas.config", EH_SASL)
+        .option("kafka.sasl.mechanism", "PLAIN")
+        .option("kafka.security.protocol", "SASL_SSL")
+        .option("kafka.request.timeout.ms", "60000")
+        .option("kafka.session.timeout.ms", "60000")
+        .option("failOnDataLoss", "false")
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:50]: provide doc for future dev on how/why to set each field

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  /**
+   * load the source data as dataframe.
+   *
+   * @return an dataframe
+   */
+  override def loadDataFrame(): DataFrame = {
+    val TOPIC = input.topics.mkString(",")
+    val BOOTSTRAP_SERVERS = input.brokers.mkString(",")
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:60]: lower case val

Comment (hangfei) [reply src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:60]: doc why do we makestring.

File: src/main/scala/com/linkedin/feathr/offline/source/accessor/StreamDataSourceAccessor.scala [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    extends DataSourceAccessor(source) {
+
+  /**
+   * get source data as a dataframe
+   *
+   * @return the dataframe
+   */
+  override def get(): DataFrame = {
+    if (!source.location.isInstanceOf[KafkaEndpoint]) {
+      throw new FeathrException(s"${source.location} is not a Kakfa streaming source location.")
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/source/accessor/StreamDataSourceAccessor.scala:28]: error message: ${source.location} is not a Kakfa streaming source location. Only Kafka source is supported right now.

File: src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.spark.sql.functions.{col, udf}
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Row, SaveMode, SparkSession}
+
+import java.nio.ByteBuffer
+import scala.collection.JavaConverters._
+import scala.collection.convert.wrapAll._
+
+class StreamingFeatureGenerator {
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala:29]: doc

File: src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+    assert(featureGenSpec.getOutputProcessorConfigs.size == 1, "Only one output sink is supported in streaming mode")
+    val outputConfig = featureGenSpec.getOutputProcessorConfigs.head
+    val timeoutMs = if (!outputConfig.getParams.hasPath("timeoutMs")) {
+      Long.MaxValue
+    } else {
+      outputConfig.getParams.getNumber("timeoutMs").longValue()
+    }
+    // Load the raw streaming source data
+    val anchorDFRDDMap = anchorToDataFrameMapper.getAnchorDFMapForGen(ss, anchors, None, false, true)
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala:46]: anchorDfRddMap

File: src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        val reader = new GenericDatumReader[GenericRecord](avroSchema)
+        val record = reader.read(null, decoder)
+        for (field <- record.getSchema.getFields) {
+          var value = record.get(field.name)
+          var fieldType = field.schema.getType
+          if (fieldType.equals(Type.UNION)) fieldType = field.schema.getTypes.get(1).getType
+          // Avro returns Utf8s for strings, which Spark SQL doesn't know how to use
+          if (fieldType.equals(Type.STRING) && value != null) value = value.toString
+          // Avro returns binary as a ByteBuffer, but Spark SQL wants a byte[]
+          if (fieldType.equals(Type.BYTES) && value != null) value = value.asInstanceOf[ByteBuffer].array
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala:64]: why not case switch here

File: src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        new CustomGenericRowWithSchema(
+          values.asScala.toArray, schemaStruct
+        )
+      }
+      val convertUDF = udf(rowForRecord)
+
+      // Streaming processing each source
+      accessor.get().writeStream
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala:75]: nit: accessor -> dfAccessor

**Comment by xiaoyongzhu** [comment]
Another bug while testing: We should make this kafka thing optional if the config does not have a kafka section, it will fail:
```
    self.kafka_endpoint = envutils.get_environment_variable_with_default(
  File "c:\users\xiaoyzhu\documents\github\feathr\feathr_project\feathr\_envvariableutil.py", line 43, in get_environment_variable_with_default
    yaml_layer = yaml_layer[arg]
KeyError: 'kafka'
```

File: src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+          val withKeyColumnDF = keyExtractor.appendKeyColumns(rowDF)
+          // Apply feature transformation
+          val transformedResult = DataFrameBasedSqlEvaluator.transform(anchor.featureAnchor.extractor.asInstanceOf[SimpleAnchorExtractorSpark],
+            withKeyColumnDF, featureNamePrefixPairs, anchor)
+          val outputJoinKeyColumnNames = getFeatureJoinKey(keyExtractor, withKeyColumnDF)
+          val selectedColumns = outputJoinKeyColumnNames ++ anchor.selectedFeatures.filter(keyTaggedFeatures.map(_.featureName).contains(_))
+          val cleanedDF = transformedResult.df.select(selectedColumns.head, selectedColumns.tail:_*)
+          val keyColumnNames = FeatureTransformation.getStandardizedKeyNames(outputJoinKeyColumnNames.size)
+          val resultFDS: DataFrame = PostGenPruner().standardizeColumns(outputJoinKeyColumnNames, keyColumnNames, cleanedDF)
+          val tableParam = "table_name"
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/generation/StreamingFeatureGenerator.scala:100]: this variable is a bit redudant. either make it constant or just use it.

File: src/main/scala/com/linkedin/feathr/offline/config/location/KafkaEndpoint.scala [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+case class KafkaSchema(@JsonProperty("type") `type`: String,
+                       @JsonProperty("avroJson") avroJson: String)
+
+@CaseClassDeserialize()
+case class KafkaEndpoint(@JsonProperty("brokers") brokers: List[String],
+                         @JsonProperty("topics") topics: List[String],
+                         @JsonProperty("schema") schema: KafkaSchema) extends InputLocation {
+  override def loadDf(ss: SparkSession, dataIOParameters: Map[String, String] = Map()): DataFrame = ???
+
+  override def getPath: String = "kafka://" + brokers.mkString(",")+":"+topics.mkString(",")
```
Comment (hangfei) [inline src/main/scala/com/linkedin/feathr/offline/config/location/KafkaEndpoint.scala:18]: document with example

**Comment by hangfei** [comment]
Discussed offline. Ideally we want users to define features without schema as well so it's aligned with offline and online. But seems Kafka doesn't provide such capability by itself yet. So instead, we will ask users to provide a schema.

In the future, if there is interface to allow kafka read without schema, we can make the schema optional.

**Comment by hangfei** [comment]
Could you add a dev guide as well for future devs?

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala (hunk 9 above)
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala:50]: These are default settings, we can expose them in the future if required

File: README.md (hunk 1 above)
Comment (xiaoyongzhu) [reply README.md:202]: Maybe in a separate PR? we should also add Feathr highlights

**Comment by hangfei** [comment]
How do we support other streaming sources, like Flink?

Maybe we should do: 

```
stream_source = StreamingSource(name="kafkaStreamingSource",
                              kafkaConfig=KafkaConfig(brokers=["feathrazureci.servicebus.windows.net:9093"],
                                                      topics=["feathrcieventhub"],
                                                      schema=schema)
                              )
```

File: feathr_project/test/test_fixture.py [hunk 18]
```diff
@@ -112,18 +114,49 @@ def snowflake_test_setup(config_path: str):
...
-
-
-
     client.build_features(anchor_list=[snowflakeFeatures])
-
     return client
 
+def kafka_test_setup(config_path: str):
+    client = FeathrClient(config_path=config_path)
+    schema = AvroJsonSchema(schemaStr="""
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/test/test_fixture.py:122]: Users will need to get it from the source data/metadata.

File: docs/concepts/feathr-capabilities.md [hunk 19]
```diff
@@ -122,3 +122,46 @@ user_item_similarity = DerivedFeature(name="user_item_similarity",
...
+### Define streaming features
+
+```python
+# Define input data schema
+schema = AvroJsonSchema(schemaStr="""
+{
+    "type": "record",
+    "name": "DriverTrips",
+    "fields": [
+        {"name": "driver_id", "type": "long"},
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply docs/concepts/feathr-capabilities.md:135]: NYC taxi schema is a too big. This is a minimum example. I think we can leave the NYC case to the example repo.

File: feathr_project/feathr/client.py [hunk 20]
```diff
@@ -117,6 +117,10 @@ def __init__(self, config_path:str = "./feathr_config.yaml", local_workspace_dir
         self.s3_endpoint = envutils.get_environment_variable_with_default(
             'offline_store', 's3', 's3_endpoint')
 
+        # Kafka configs
+        self.kafka_endpoint = envutils.get_environment_variable_with_default(
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/feathr/client.py:121]: Updated

File: docs/concepts/feature-generation.md [hunk 21]
```diff
@@ -44,3 +44,62 @@ res = client.get_online_features('nycTaxiDemoFeature', '265', [
 ```
 
 After we finish running the materialization job, we can get the online features by querying the feature name, with the corresponding keys. In the exmaple above, we query the online features called `f_location_avg_fare` and `f_location_max_fare`, and query with a key `265` (which is the location ID).
+
+## Streaming feature ingestion
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply docs/concepts/feature-generation.md:48]: Will do in separate PR

File: docs/concepts/feature-generation.md [hunk 22]
```diff
@@ -44,3 +44,62 @@ res = client.get_online_features('nycTaxiDemoFeature', '265', [
...
+```python
+driver_id = TypedKey(key_column="driver_id",
+                     key_column_type=ValueType.INT64,
+                     description="driver id",
+                     full_name="nyc driver id")
+
+kafkaAnchor = FeatureAnchor(name="kafkaAnchor",
+                            source=stream_source,
+                            features=[Feature(name="f_modified_streaming_count",
+                                              feature_type=INT32,
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply docs/concepts/feature-generation.md:86]: Added

File: src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala (hunk 3 above)
Comment (jaymo001) [reply src/test/scala/com/linkedin/feathr/offline/FeatureGenIntegTest.scala]: HOCON file does not support it

File: feathr_project/feathrcli/data/feathr_user_workspace/feathr_config.yaml [hunk 23]
```diff
@@ -63,9 +63,9 @@ offline_store:
     jdbc_database: "feathrtestdb"
     jdbc_table: "feathrtesttable"
 
-# reading from streaming source is coming soon
-# streaming_source:
-#   kafka_connection_string: ""
+  # reading from streaming source
+  kafka:
+    kafka_endpoint: 'sb://feathrazureci.servicebus.windows.net/'
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/feathrcli/data/feathr_user_workspace/feathr_config.yaml]: Done

File: src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+
+  private def getKafkaAuth(ss: SparkSession): String = {
+    // If user set password, then we use password to auth
+    ss.conf.getOption(kafkaEndpoint) match {
+      case Some(_) =>
+          val accessKeyName = ss.conf.get(sharedAccessKeyName)
+          val endpoint = ss.conf.get(kafkaEndpoint)
+          val username = ss.conf.get(kafkaUsername)
+          val accessKey = ss.conf.get(sharedAccessKey)
+          val EH_SASL = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\""+username+"\"" +
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/offline/source/dataloader/stream/KafkaDataLoader.scala]: As discussed offline, will leave to future work.

File: feathr_project/feathr/_envvariableutil.py [hunk 25]
```diff
@@ -42,6 +42,8 @@ def get_environment_variable_with_default(self, *args):
                 for arg in args:
                     yaml_layer = yaml_layer[arg]
                 return yaml_layer
+            except KeyError as exc:
+                return ""
             except yaml.YAMLError as exc:
```
Comment (xiaoyongzhu) [inline feathr_project/feathr/_envvariableutil.py:47]: Can we add logger here to indicate that it's not found?

**Review by hangfei — DISMISSED** [review]

**Comment by hangfei** [comment]
> How do we support other streaming sources, like Flink?
> 
> Maybe we should do:
> 
> ```
> stream_source = StreamingSource(name="kafkaStreamingSource",
>                               kafkaConfig=KafkaConfig(brokers=["feathrazureci.servicebus.windows.net:9093"],
>                                                       topics=["feathrcieventhub"],
>                                                       schema=schema)
>                               )
> ```

Discussed offline. The current syntax just work so no need to introduce anohter layer.

File: feathr_project/test/test_azure_kafka_e2e.py [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+    test_workspace_dir = Path(__file__).parent.resolve() / "test_user_workspace"
+
+    client = kafka_test_setup(os.path.join(test_workspace_dir, "feathr_config.yaml"))
+    redisSink = RedisSink(table_name="kafkaSampleDemoFeature", streaming=True, streamingTimeoutMs=10000)
+    settings = MaterializationSettings(name="kafkaSampleDemo",
+                                   sinks=[redisSink],
+                                   feature_names=['f_modified_streaming_count']
+                                   )
+    client.materialize_features(settings)
+    client.wait_job_to_finish(timeout_sec=600)
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/test/test_azure_kafka_e2e.py:21]: We set a timeout of 10 second already, so this should not be needed in the test.

File: feathr_project/feathrcli/data/feathr_user_workspace/feathr_config.yaml [hunk 27]
```diff
@@ -63,9 +63,10 @@ offline_store:
     jdbc_database: "feathrtestdb"
     jdbc_table: "feathrtesttable"
 
-# reading from streaming source is coming soon
-# streaming_source:
-#   kafka_connection_string: ""
+  # reading from streaming source
+streaming:
+  kafka:
+    kafka_endpoint: 'sb://feathrazureci.servicebus.windows.net/'
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply feathr_project/feathrcli/data/feathr_user_workspace/feathr_config.yaml]: Removed

**Review by hangfei — APPROVED** [review]

**Review by xiaoyongzhu — APPROVED** [review]

**Pull request merged**